client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java

```diff
@@ -1508,6 +1511,32 @@ private TxnSegments convert(CreateTxnResponse response) {
                 });
     }
 
+    @Override
+    public CompletableFuture<List<UUID>> listCompletedTxns(Stream stream) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        Preconditions.checkNotNull(stream, "stream");
+        final long requestId = requestIdGenerator.get();
+        long traceId = LoggerHelpers.traceEnter(log, "listCompletedTxns", stream, requestId);
+
+        final CompletableFuture<ListCompletedTxnResponse> result = this.retryConfig.runAsync(() -> {
+            RPCAsyncCallback<ListCompletedTxnResponse> callback = new RPCAsyncCallback<>(traceId, "listCompletedTransactions", stream);
+            new ControllerClientTagger(client, timeoutMillis).withTag(requestId, LIST_COMPLETED_TRANSACTIONS,
+                            stream.getScope(), stream.getStreamName())
+                    .listCompletedTransactions(ListCompletedTxnRequest.newBuilder()
+                                    .setStreamInfo(ModelHelper.createStreamInfo(stream.getScope(), stream.getStreamName()))
+                                    .build(),
+                            callback);
+            return callback.getFuture();
+        }, this.executor);
+        return result.thenApplyAsync(x -> x.getTxnIdList().stream().map(uuid -> encode(uuid)).collect(Collectors.toList()), this.executor)
```

Comment:
would be good if you rename `x` with actual typeName

Reply:
Done as suggested.